File: Sources/KillerSudokuCore/Coordinate.swift
public struct Coordinate: Hashable, Sendable, Codable {
    public let row: Int
    public let column: Int

    public init(row: Int, column: Int) {
        precondition((0..<9).contains(row) && (0..<9).contains(column), "Coordinate out of range")
        self.row = row
        self.column = column
    }

    public var boxIndex: Int { (row / 3) * 3 + (column / 3) }

    /// Every cell on the board, once each. Shared by anything that needs to enumerate or
    /// partition the whole 9x9 grid.
    public static let all: [Coordinate] = {
        var coordinates: [Coordinate] = []
        for row in 0..<9 {
            for column in 0..<9 {
                coordinates.append(Coordinate(row: row, column: column))
            }
        }
        return coordinates
    }()
}

File: desktop-kotlin/src/main/kotlin/me/gileskayo/killersudoku/core/Coordinate.kt
package me.gileskayo.killersudoku.core

import kotlinx.serialization.Serializable

@Serializable
data class Coordinate(val row: Int, val column: Int) {
    init {
        require(row in 0..8 && column in 0..8) { "Coordinate out of range" }
    }

    val boxIndex: Int get() = (row / 3) * 3 + (column / 3)

    companion object {
        /** Every cell on the board, once each. Shared by anything that needs to enumerate or
         * partition the whole 9x9 grid. */
        val all: List<Coordinate> = buildList {
            for (row in 0..8) {
                for (column in 0..8) {
                    add(Coordinate(row, column))
                }
            }
        }
    }
}
